hamt/hamt.go

```diff
@@ -39,13 +39,28 @@ const (
 	HashMurmur3 uint64 = 0x22
 )
 
+type nodeType int
+
+const (
+	invalidNode nodeType = iota
+	shardNode
+	shardValueNode
+)
+
+func (ds *Shard) nodeType() nodeType {
```

Comment:
Yeeees!! Definitely that is conceptually much clearer, since we only care about knowing where does our directory stop and the rest of the DAG hierarchy continues, that makes much sense.

Reply:
My main concern with the "node" terminology here is that, although conceptually from the UnixFS layer perspective this is another (separate) node (a leaf node in the trie), actually it's being encoded inside its parent (internal) node in the DAG layer (i.e., that node is not stored separately in the DAG service). But I guess that is a correct abstraction, this collapse at the DAG level (for performance reasons) doesn't need to be complicating the logic here.